### ISSUE TITLE
Extra elements created in multi-dimensional arrays

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8866]
Migrated from [rt.perl.org#42363](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=42363) (status was 'rejected')

Searchable as RT42363$

**Comment by p5pRT** [comment]
### From andyt@andy-t.org

#### Created by andyt@andy-t.org



While trying to write a program that makes use of a two\-dimensional
array \(list of lists\, to be exact\)\, and trying to debug it\, I ran into
a possible bug\.  I was trying to access a non\-existing element\, outside
of array boundaries\, and this element\, as well as all intermediate
elements\, were created for me\.  Needless to say\, they were all empty\.
Here is a simple piece of code to reproduce the problem&#8203;:

@&#8203;a=\(\[1\]\,\[2\]\);
print $a\[8\]\[11\];         \# Returns empty string
print "$\#a\\n";           \# Returns 8
print "$\#\{$a\[8\]\}\\n";     \# Returns \-1
print "$\#\{$a\[20\]\}\\n";    \# Returns \-1
print "$\#a\\n";           \# Returns 20

This does not happen with a simple list\.  Since \`print' is not an 
assignment\, I do not see why it should create new elements\.  I wonder
if this behavior is unique to \`print' function\.\.\.

Regards\,

Andy

Dr Andy Tsouladze
Sr Unix SysAdmin
United Airlines

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.8.7:

Configured by root at Wed Jul 13 12:01:19 PDT 2005.

Summary of my perl5 (revision 5 version 8 subversion 7) configuration:
  Platform:
    osname=linux, osvers=2.4.31, archname=i486-linux
    uname='linux midas 2.4.31 #6 sun jun 5 19:04:47 pdt 2005 i686 unknown unknown gnulinux '
    config_args='-de -Dprefix=/usr -Dcccdlflags=-fPIC -Dinstallprefix=/usr -Doptimize=-O2 -march=i486 -mcpu=i686 -Dinc_version_list=5.8.6 5.8.5 5.8.4 5.8.3 5.8.2 5.8.1 5.8.0 -Darchname=i486-linux'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -march=i486 -mcpu=i686',
    cppflags='-fno-strict-aliasing -pipe -I/usr/local/include'
    ccversion='', gccversion='3.3.6', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.3.5.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.3.5'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.7:
    /usr/lib/perl5/5.8.7/i486-linux
    /usr/lib/perl5/5.8.7
    /usr/lib/perl5/site_perl/5.8.7/i486-linux
    /usr/lib/perl5/site_perl/5.8.7
    /usr/lib/perl5/site_perl
    .


Environment for perl v5.8.7:
    HOME=/home/andyt
    LANG=en_US
    LANGUAGE (unset)
    LC_ALL=POSIX
    LC_COLLATE=C
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/local/bin:/opt/kde/bin:/bin:/usr/bin:/usr/X11R6/bin:/usr/openwin/bin:/usr/games:/usr/share/texmf/bin:/etc:/usr/sbin:/usr/bin/X11:/home/andyt/bin
    PERL_BADLANG (unset)
    SHELL=/bin/tcsh

```
</details>

**Comment by p5pRT** [comment]
### From shouldbedomo@mac.com

On 2007–04–09\, at 07&#8203;:22\, andyt@&#8203;andy\-t\.org \(via RT\) wrote&#8203;:
> While trying to write a program that makes use of a two\-dimensional
> array \(list of lists\, to be exact\)\, and trying to debug it\, I ran into
> a possible bug\.  I was trying to access a non\-existing element\,  
> outside
> of array boundaries\, and this element\, as well as all intermediate
> elements\, were created for me\.

It's not a bug; Perl's supposed to work this way&#8203;: if you say you want  
to access a particular array element\, Perl does its best to arrange  
things so that you can\, even if that element does not exist\. It's  
documented in passing in a number of places\, under the name of  
"autovivification" \(which you would never have guessed in a month of  
Sundays\)\. See\, for example\, the perldata man page and\, if you can  
easily look at Perl 5\.8\.8 or the third edition of "Programming Perl"\,  
perlglossary\. \(You filed your report using 5\.8\.7\.\)

Autovivification's generally useful as an aspect of Perl's "do what I  
mean" philosophy\, but if you don't want Perl to work this way \-\- as  
you might not if used to Java\, for example \-\- you'll have to write  
array\-bound checks yourself&#8203;: compare the proposed index against  
$\#whatever \-\- the index of currently last element of the array \-\-  
before using it\.

For example\,

$ perl \-lwe '@&#8203;a = \(\[1\, 2\]\, \[3\, 4\, 5\]\); print $\#a; for \(0\, 3\) \{exists  
$a\[$\_\]\[0\]\}; print $\#a'
1
3
$ perl \-lwe '@&#8203;a = \(\[1\, 2\]\, \[3\, 4\, 5\]\); print $\#a; for \(0\, 3\) \{exists  
$a\[$\_\]\[0\] unless $\_ > $\#a\}; print $\#a'
1
1

In the first case\, the outer array got extended as the result of a  
test of whether an inner array element existed; in the second\, the  
bounds check prevents this from happening\.
\-\- 
Dominic Dunlop

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From guest@guest.guest.xxxxxxxx

> It's not a bug; Perl's supposed to work this way&#8203;: if you say you 
want  
> to access a particular array element\, Perl does its best to arrange  
> things so that you can\, even if that element does not exist\. It's  
> documented in passing in a number of places\, under the name of  
> "autovivification" \(which you would never have guessed in a month of  
> Sundays\)\.

I do know all this\, including autovivification \(although I admit I 
forgot the word\)\. However\, I do think that \`print' should not change 
anything\.  If I assign a value to a specific element\, yes\, this should 
and does work\, as in

perl \-le '@&#8203;a=\(1\); $a\[5\]=2; print $\#a'
5

Note that in case of this 1\-D list\, \`print' would do nothing&#8203;:

perl \-le '@&#8203;a=\(1\); print $a\[5\]; print $\#a'

0

But if I try to access the same 1\-D array as 2\-D\, \`print' suddenly 
behaves differently\.

perl \-le '@&#8203;a=\(1\); print $a\[7\]\[5\]; print $\#a; print $a\[7\];'

7
ARRAY\(0x812fc18\)

Now we have a 2\-D array\, and if\, without realizing it\, we try to loop 
through @&#8203;a\, we are in for a few surprises \- empty elements 1\.\.6\, and an 
array for 7\.

In short\, such behavior may not constitute a bug\, but I still believe 
it is undesirable because non\-assigning functions like \`print' should 
not be able to modify values\.  In my mind\, it contradicts the "do what 
I mean" principle\.  Besides\, even this is inconsistent \(the way I see 
it\, of course ;\)  \)\.

Regards\,

Andy

**Comment by p5pRT** [comment]
### From p5p@spam.wizbit.be

On Mon Apr 09 17&#8203;:02&#8203;:05 2007\, guest wrote&#8203;:
> > It's not a bug; Perl's supposed to work this way&#8203;: if you say you 
> want  
> > to access a particular array element\, Perl does its best to 
arrange  
> > things so that you can\, even if that element does not exist\. It's  
> > documented in passing in a number of places\, under the name of  
> > "autovivification" \(which you would never have guessed in a month 
of  
> > Sundays\)\.
> 
> I do know all this\, including autovivification \(although I admit I 
> forgot the word\)\. However\, I do think that \`print' should not change 
> anything\.  If I assign a value to a specific element\, yes\, this 
should 
> and does work\, as in
> 
> perl \-le '@&#8203;a=\(1\); $a\[5\]=2; print $\#a'
> 5

This is not autovivification\.
http&#8203;://en\.wikipedia\.org/wiki/Autovivification

Autovivification is the automatic creation of a reference when an 
undefined value is dereferenced\.

$a\[15\]\[2\] = 3;  and  print $a\[25\]\[4\]; would be autovivfication\.
You are trying to derefence $a\[15\] in order to get to element \[2\]\.


Kind regards\,

Bram

**Comment by p5pRT** [comment]
p5p@spam.wizbit.be - Status changed from 'open' to 'rejected'
